RavenDB-11310 Server Tx Merger

ppekrol opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-11310

### Type of change

- Optimization

### How risky is the change?

- High

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

ppekrol committed:
RavenDB-11310 moving files around

ppekrol committed:
RavenDB-11310 working on tx merger abstraction

ppekrol committed:
RavenDB-11310 working on tx merger abstraction

ppekrol committed:
RavenDB-11310 working on tx merger abstraction

ppekrol commented:
Nothing to see here at this point, just created abstraction for future work and checking CI.

ppekrol committed:
RavenDB-11310 fixed initialization order

ppekrol committed:
RavenDB-11310 fixing NRE

ppekrol committed:
RavenDB-11310 implemented LeaderApplyCommand

ppekrol committed:
RavenDB-11310 implemented SwitchToCandidateStateCommand

ppekrol committed:
RavenDB-11310 implemented ElectorCastVoteInTermCommand and ElectorCastVoteInTermWithShouldGrantVoteCommand

ppekrol committed:
RavenDB-11310 implemented CandidateCastVoteInTermCommand and FollowerApplyCommand

ppekrol committed:
RavenDB-11310 implemented LeaderTruncateLogCommand

ayende commented on src/Raven.Server/Rachis/Candidate.cs:
```diff
@@ -61,7 +66,7 @@ private void Run()
                     if (clusterTopology.Members.Count == 1)
                     {
-                        CastVoteForSelf(ElectionTerm + 1, "Single member cluster, natural leader");
+                        await CastVoteForSelfAsync(ElectionTerm + 1, "Single member cluster, natural leader");
```
Not happy with this. 
It violates an important consideration for us, that we can always see the operations of the leader/ candidate / follower in their thread.
This way, we may return to _another_ thread here, and the candidate thread is just stalled.

ayende commented on src/Raven.Server/Rachis/RachisConsensus.cs:
```diff
@@ -440,6 +444,9 @@ public unsafe void Initialize(StorageEnvironment env, RavenConfiguration configu
                 DebuggerAttachedTimeout.LongTimespanIfDebugging(ref _electionTimeout);
 
                 ContextPool = new ClusterContextPool(changes, _persistentState, configuration.Memory.MaxContextSizeToKeep);
+                TxMerger = new ClusterTransactionOperationsMerger(this, configuration, time, shutdown);
```
we have some "local" operations that are running on the server store, they are also going through here (local cert, for example)

ppekrol replied:
This will be problematic, since they are not using ClusterOperationContext, yet same store. I've reviewed them and most of those operations are ultra-rare. Probably we do not need to apply tx merger there. Now I'm focusing only on all of the operations bound to ClusterOperationContext.

ppekrol committed:
RavenDB-11310 we cannot use awaits because it will switch thread and we are setting the priority of the thread

ppekrol committed:
RavenDB-11310 bumping leader thread priority to above normal

ppekrol committed:
RavenDB-11310 introduced ThreadGuardian

ppekrol committed:
RavenDB-11310 fixes

ppekrol committed:
RavenDB-11310
- dispose command blittable as soon as possible
- for HasHistoryLog we do not need blittable

ppekrol committed:
RavenDB-11310 implemented LeaderEmptyQueueCommand

ppekrol committed:
RavenDB-11310 cleanup

ppekrol committed:
RavenDB-11310 unwrapping exception

ppekrol committed:
RavenDB-11310 added size threshold for amount of commands to process

ppekrol committed:
RavenDB-11310 added maxCommandsToProcessTime

ppekrol committed:
RavenDB-11310 implemented LeaderModifyTopologyCommand

ppekrol committed:
RavenDB-11310 renames

ppekrol committed:
RavenDB-11310 implemented CastVoteInTermCommand

ppekrol committed:
RavenDB-11310 implemented UpdateNodeTagCommand

ppekrol committed:
RavenDB-11310 fixing double apply issue with commands

ppekrol committed:
RavenDB-11310 fixing ModifyTopologyAsync

ppekrol committed:
RavenDB-11310 pre-validate in read transaction

ppekrol committed:
RavenDB-11310 implemented FollowerReadAndCommitSnapshotCommand

ppekrol commented:
New PR: https://github.com/ravendb/ravendb/pull/15686
